File: src/research_artifact_uploader/records.py
from __future__ import annotations

from pathlib import Path
import json


def _write_parent(path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)


def append_jsonl(records_path: Path, payload: dict) -> None:
    _write_parent(records_path)
    with records_path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(payload, ensure_ascii=False) + "\\n")


def _human_size(size_bytes: int) -> str:
    gib = size_bytes / (1024 ** 3)
    if gib >= 1:
        return f"{gib:.1f} GiB"
    mib = size_bytes / (1024 ** 2)
    return f"{mib:.1f} MiB"


def append_markdown(records_path: Path, payload: dict) -> None:
    _write_parent(records_path)
    exists = records_path.exists()
    with records_path.open("a", encoding="utf-8") as f:
        if not exists:
            f.write("# Upload Records\\n\\n")
            f.write("| Time | Run | Project | Size | Speed | Remote URI | SHA256 |\\n")
            f.write("| --- | --- | --- | ---: | ---: | --- | --- |\\n")
        speed = payload.get("avg_mib_s")
        speed_display = "n/a" if speed is None else f"{speed:.2f}"
        f.write(
            f"| {payload['time']} | {payload['run_name']} | {payload['project']} | "
            f"{_human_size(payload['size_bytes'])} | {speed_display} MiB/s | "
            f"`{payload['archive_uri']}` | `{payload['sha256']}` |\\n"
        )


def read_records(records_path: Path, last: int = 10):
    if not records_path.exists():
        return []
    lines = records_path.read_text(encoding="utf-8").splitlines()
    rows = []
    for line in lines[-last:]:
        if not line.strip():
            continue
        try:
            rows.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return rows

File: src/records.rs
use std::{
    fs::{self, OpenOptions},
    io::Write,
    path::Path,
};

use serde_json::Value;

use crate::AppResult;

#[derive(Debug)]
pub struct RecordRow {
    pub time: String,
    pub run_id: String,
    pub run_name: String,
    pub project: String,
    pub archive_uri: String,
    pub size_bytes: u64,
    pub avg_mib_s: Option<f64>,
}

pub fn append_jsonl(path: &Path, payload: &Value) -> AppResult<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("create records directory failed: {e}"))?;
    }

    let text = format!("{}\n", payload);
    OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .map_err(|e| format!("open records jsonl failed: {e}"))?
        .write_all(text.as_bytes())
        .map_err(|e| format!("write records jsonl failed: {e}"))?;

    Ok(())
}

pub fn append_markdown(path: &Path, payload: &Value) -> AppResult<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("create records directory failed: {e}"))?;
    }

    let exists = path.exists();
    let mut fh = OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .map_err(|e| format!("open records markdown failed: {e}"))?;

    if !exists {
        let header = "# Upload Records\n\n| Time | Run | Project | Size | Speed | Remote URI | SHA256 |\n| --- | --- | --- | ---: | ---: | --- | --- |\n";
        fh.write_all(header.as_bytes())
            .map_err(|e| format!("write markdown header failed: {e}"))?;
    }

    let speed = payload
        .get("avg_mib_s")
        .and_then(|x| x.as_f64())
        .unwrap_or(0.0);
    let size = payload
        .get("size_bytes")
        .and_then(|x| x.as_u64())
        .unwrap_or(0);
    let row = format!(
        "| {} | {} | {} | {:.1} MiB | {:.2} MiB/s | `{}` | `{}` |\n",
        payload.get("time").and_then(|x| x.as_str()).unwrap_or(""),
        payload.get("run_name").and_then(|x| x.as_str()).unwrap_or(""),
        payload.get("project").and_then(|x| x.as_str()).unwrap_or(""),
        bytes_to_mib(size),
        speed,
        payload
            .get("archive_uri")
            .and_then(|x| x.as_str())
            .unwrap_or(""),
        payload
            .get("sha256")
            .and_then(|x| x.as_str())
            .unwrap_or(""),
    );
    fh.write_all(row.as_bytes())
        .map_err(|e| format!("write markdown row failed: {e}"))?;
    Ok(())
}

pub fn read_records(path: &Path, last: usize) -> Vec<RecordRow> {
    let content = match fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return vec![],
    };

    let mut rows = Vec::new();
    for line in content.lines().rev().take(last) {
        if line.trim().is_empty() {
            continue;
        }
        if let Ok(value) = serde_json::from_str::<Value>(line) {
            rows.push(RecordRow {
                time: value
                    .get("time")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                run_id: value
                    .get("run_id")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                run_name: value
                    .get("run_name")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                project: value
                    .get("project")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                archive_uri: value
                    .get("archive_uri")
                    .and_then(|x| x.as_str())
                    .unwrap_or("")
                    .to_string(),
                size_bytes: value
                    .get("size_bytes")
                    .and_then(|x| x.as_u64())
                    .unwrap_or(0),
                avg_mib_s: value.get("avg_mib_s").and_then(|x| x.as_f64()),
            });
        }
    }

    rows.reverse();
    rows
}

fn bytes_to_mib(size: u64) -> f64 {
    (size as f64) / (1024f64 * 1024f64)
}
